Bucky/adrs

ebuchman opened the pull request:
Wrote some Architecture Decision Records for 

- the recent changes to the ABCI, including some additional ones we should make. basically complete
- proposed changes to our crypto libs - still needs to be fleshed out and references added

A lot of this information is already captured across issues, especially starting from https://github.com/tendermint/tendermint/issues/1568, but in light of Ismail's HDkeys design doc, thought we should try to catch up here as well. 

We should also add ADRs for the proposed changes to:

- NextValidatorSet
- BFT Time


Note the distinction between an ADR and a spec. We should be using both for basically everything. The ADR provides the context and intuition and justification for a change in architecture, or for the architecture of something new. The spec is much more compressed and streamlined summary of everything as it stands today.

xla commented on docs/architecture/adr-009-ABCI-design.md:
```diff
@@ -0,0 +1,260 @@
+- It will not be required for PubKey encoding
+- It will not be required for computing PubKey addresses
+
+That said, we are working to make Amino a huge success, and to become Protobuf4.
```
`Protobuf4` -> `proto4`

xla commented on docs/architecture/adr-009-ABCI-design.md:
````diff
@@ -0,0 +1,260 @@
+}
+```
+
+where `type` can be:
````
If `type` is only one of two to three possibilities why are we not using an enum for increased hardening?

ebuchman replied:
We actually want `PubKey` to be an interface, where you don't necessarily know all the types that could implement it.

ValarDragon replied:
Since we have to fork anyway to add a new pubkey type, not sure I follow why this couldn't be an enum?

ebuchman replied:
This argument doesn't really hold for Tendermint, but I think it might for an app like the SDK where some module could have a custom implementation of pubkey

xla commented on docs/architecture/adr-009-ABCI-design.md:
```diff
@@ -0,0 +1,260 @@
+}
+
+message ResponseInitChain {
+    ConsensusParams consensus_params
```
Could be helpful to expand on this field and clarify why we return it back to the caller and the content is.

ebuchman replied:
`Chain ResponseInitChain to allow the app to specify the initial validator set
and consensus parameters.` is that better ? ConsensusParams definition doesnt belong here

Liamsi commented on docs/architecture/adr-010-crypto-changes.md:
```diff
@@ -0,0 +1,76 @@
+
+### Merkle Trees
+
+TODO
```
Will there be changes that will be included here? E.g., SimpleMerkle also uses the truncated sha256 now:
https://github.com/tendermint/go-crypto/blob/master/CHANGELOG.md#070

ebuchman replied:
Yes, would be great if you could help complete it. I left it for now as we were still discussing changes to the internal nodes to prevent 2nd preimages ...

ebuchman commented on docs/architecture/adr-010-crypto-changes.md:
```diff
@@ -0,0 +1,76 @@
+
+### Addresses
+
+ED25519 addresses are computed using the RIPEMD160 of the Amino encoding of the public key.
```
I doubled down on current tense since we're trying to describe the current state and then make proposals for the new state. I'll add a short README

ebuchman commented:
Addressed feedback and rebased on develop. Crypto one still needs some work but maybe @Liamsi can get to it as part of preparations for next release with go-crypto merged and any final changes to the merkle tree or other pieces. thanks!
